src/storage/src/hummock/sstable/delete_range_aggregator.rs

```diff
@@ -0,0 +1,311 @@
+    fn valid(&self) -> bool;
+}
+
+pub struct SingleDeleteRangeIterator {
+    agg: Arc<DeleteRangeAggregator>,
+    seek_idx: usize,
+}
+
+impl DeleteRangeIterator for SingleDeleteRangeIterator {
+    fn start_user_key(&self) -> &[u8] {
+        &self.agg.delete_tombstones[self.seek_idx].start_user_key
+    }
+
+    fn end_user_key(&self) -> &[u8] {
+        &self.agg.delete_tombstones[self.seek_idx].end_user_key
+    }
+
+    fn current_epoch(&self) -> HummockEpoch {
+        self.agg.delete_tombstones[self.seek_idx].sequence
+    }
+
+    fn next(&mut self) {
+        self.seek_idx += 1;
+    }
+
+    fn seek_to_first(&mut self) {
+        self.seek_idx = 0;
+    }
+
+    fn seek(&mut self, target_key: &[u8]) {
+        self.seek_idx = 0;
+        while self.seek_idx < self.agg.delete_tombstones.len()
+            && self.agg.delete_tombstones[self.seek_idx]
+                .end_user_key
+                .as_slice()
+                .le(target_key)
+        {
+            self.seek_idx += 1;
+        }
+    }
+
+    fn valid(&self) -> bool {
+        self.seek_idx < self.agg.delete_tombstones.len()
+    }
+}
+
+pub struct DeleteRangeAggregatorIterator<I: DeleteRangeIterator> {
+    inner: I,
+    end_user_key_index: BinaryHeap<SortedBoundary>,
+    epoch_index: BTreeSet<HummockEpoch>,
+    watermark: u64,
+}
+
+impl<I: DeleteRangeIterator> DeleteRangeAggregatorIterator<I> {
+    pub fn should_delete(&mut self, target_key: &[u8], epoch: HummockEpoch) -> bool {
+        if epoch >= self.watermark {
+            return false;
+        }
+
+        // take the smallest end_user_key which would never cover the current key and remove them
+        //  from covered epoch index.
+        while !self.end_user_key_index.is_empty() {
+            let item = self.end_user_key_index.peek().unwrap();
+            if item.user_key.as_slice().gt(target_key) {
+                break;
+            }
+
+            // The correctness of the algorithm needs to be guaranteed by "the epoch of the
+            // intervals covering each other must be different".
+            self.epoch_index.remove(&item.sequence);
+            self.end_user_key_index.pop();
+        }
+        while self.inner.valid() && self.inner.start_user_key().le(target_key) {
+            let sequence = self.inner.current_epoch();
+            if sequence > self.watermark || self.inner.end_user_key().le(target_key) {
+                self.inner.next();
+                continue;
+            }
+            self.end_user_key_index.push(SortedBoundary {
+                user_key: self.inner.end_user_key().to_vec(),
+                sequence,
+            });
+            self.epoch_index.insert(sequence);
+            self.inner.next();
+        }
+
+        // There may be several epoch, we only care the largest one.
+        self.epoch_index
+            .last()
+            .map(|tombstone_epoch| *tombstone_epoch >= epoch)
+            .unwrap_or(false)
+    }
+}
```

Comment:
I think `should_delete` assumes that `target_key` is passed in key order during the lifetime of DeleteRangeAggregatorIterator. We should document this assumption.

Reply:
🤔  I have add some comment